File: app/src/main/java/com/minerdev/greformanager/ImageRepository.java
package com.minerdev.greformanager;

import android.app.Application;

import androidx.lifecycle.LiveData;

import java.sql.Timestamp;
import java.util.List;

public class ImageRepository {
    private ImageDao imageDao;
    private LiveData<List<Image>> allImages;

    public ImageRepository(Application application) {
        GreDatabase db = GreDatabase.getDatabase(application);
        imageDao = db.imageDao();
        allImages = imageDao.getAllImages();
    }

    LiveData<Timestamp> getLatestUpdatedAt(int house_id) {
        return imageDao.getLatestUpdatedAt(house_id);
    }

    public LiveData<List<Image>> getAllImages() {
        return allImages;
    }

    public LiveData<List<Image>> getOrderByPosition(int house_id) {
        return imageDao.getImages(house_id);
    }

    public void insert(Image image) {
        GreDatabase.databaseWriteExecutor.execute(() -> imageDao.insert(image));
    }

    public void insert(Image... images) {
        GreDatabase.databaseWriteExecutor.execute(() -> imageDao.insert(images));
    }

    public void updateOrInsert(Image image) {
        GreDatabase.databaseWriteExecutor.execute(() -> {
            Image result = imageDao.getImage(image.house_id, image.position);
            if (result == null) {
                imageDao.insert(image);

            } else {
                imageDao.update(image);
            }
        });
    }

    public void deleteAll(int house_id) {
        GreDatabase.databaseWriteExecutor.execute(() -> imageDao.deleteAll(house_id));
    }

    public void deleteAll() {
        GreDatabase.databaseWriteExecutor.execute(() -> imageDao.deleteAll());
    }
}

File: app/src/main/java/com/minerdev/greformanager/ImageRepository.kt
package com.minerdev.greformanager

import android.app.Application
import androidx.lifecycle.LiveData

class ImageRepository(application: Application) {
    private val imageDao: ImageDao?
    val allImages: LiveData<List<Image>>?
    fun getLastUpdatedAt(house_id: Int): Long? {
        return imageDao!!.getLastUpdatedAt(house_id)
    }

    fun getOrderByPosition(house_id: Int): LiveData<List<Image>> {
        return imageDao!!.getImages(house_id)
    }

    fun insert(image: Image) {
        GreDatabase.databaseWriteExecutor.execute { imageDao!!.insert(image) }
    }

    fun insert(images: List<Image>) {
        GreDatabase.databaseWriteExecutor.execute { imageDao!!.insert(images) }
    }

    fun updateOrInsert(image: Image) {
        GreDatabase.run {
            databaseWriteExecutor.execute {
                val result = imageDao!!.getImage(image.house_id, image.position.toInt())
                if (result == null) {
                    imageDao.insert(image)
                } else {
                    imageDao.update(image)
                }
            }
        }
    }

    fun deleteAll(house_id: Int) {
        GreDatabase.databaseWriteExecutor.execute { imageDao!!.deleteAll(house_id) }
    }

    fun deleteAll() {
        GreDatabase.databaseWriteExecutor.execute { imageDao!!.deleteAll() }
    }

    init {
        val db: GreDatabase? = GreDatabase.getDatabase(application)
        imageDao = db!!.imageDao()
        allImages = imageDao.allImages
    }
}